notification/src/notification-config/notificationConfig.ts

```diff
@@ -1,57 +1,98 @@
-import { Entity, Column, PrimaryGeneratedColumn } from 'typeorm'
+import { hasProperty, isObject } from '../validators'
 
-export enum CONFIG_TYPE{
-  WEBHOOK = 'webhook',
-  SLACK = 'slack',
-  FCM = 'fcm'
+export enum NotificationType {
+  WEBHOOK = 'WEBHOOK',
+  SLACK = 'SLACK',
+  FCM = 'FCM'
 }
 
-export class NotificationConfig {
-  @Column()
-  pipelineId!: number
-
-  @Column()
-  condition!: string
+export interface NotificationConfig {
+  id: number
+  pipelineId: number
+  condition: string
+  type: NotificationType
+  parameter: NotificationParameter
 }
```

Comment:
How about changing the model to this, because your current model does not force consistency between `type` and `parameter` (e.g. a NotificationConfig can have a slack notification type but webhook parameters)? With my suggestion, also some typecasts will be unnecessary and can be removed. 

```suggestion
export type NotificationConfig = SlackNotification | WebhookNotification | FirebaseNotification

export interface BaseNotification {
  id: number
  pipelineId: number
  condition: string
}

export interface SlackNotification extends BaseNotification{
  type: NotificationType.SLACK
  parameter: SlackParameter
}

export interface WebhookNotification extends BaseNotification{
  type: NotificationType.WEBHOOK
  parameter: WebhookParameter
}

export interface FirebaseNotification extends BaseNotification{
  type: NotificationType.FCM
  parameter: FirebaseParameter
}
```

Reply:
Good point! I'm currently struggling to make it pretty :/ 
Another solution is here, which is much to complex IMHO: https://stackoverflow.com/questions/56949513/typescript-type-of-a-property-dependent-on-another-property-within-the-same-obj

There are a few switch-case statements in the code about the type and linked behavior. I'd like to get rid of that... probably by using objects? Ah it would be so nice to be able to overload methods with different subtypes and the compiler chooses the right one to execute...

Reply:
The other solutions from StackOverflow are actually the same base idea, just with fancier TypeScript syntax.

Well, that's the way of doing it in a functional programming language. But what is not pretty in your opinion? I have just played around a bit and came up with the following: https://github.com/jsone-studios/open-data-service/commit/8213019250b7aad341ebf8b138a753f179b2da85

Reply:
Well, it bugs me that we have multiple places in the code with switch-case statements that you have to go over when adding e.g. another notification config. But nevermind, it's not that big of a deal.

I tried to simplify some other stuff here and there and even got rid of one switch-case in the database repository.